handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -881,98 +916,74 @@ public final SSLEngineResult unwrap(
-                            // We read everything return now.
-                            return newResult(isInboundDone() ? CLOSED : OK, bytesConsumed, bytesProduced, status);
+                            // No space left in the destination buffer, skip it.
+                            dstsOffset++;
+                            continue;
                         }
-                    } else {
-                        int sslError = SSL.getError(ssl, bytesRead);
-                        switch (sslError) {
-                            case SSL.SSL_ERROR_ZERO_RETURN:
-                                // This means the connection was shutdown correctly, close inbound and outbound
-                                if (!receivedShutdown) {
-                                    closeAll();
-                                }
-                                // fall-trough!
-                            case SSL.SSL_ERROR_WANT_READ:
-                            case SSL.SSL_ERROR_WANT_WRITE:
-                                // break to the outer loop
-                                return newResult(isInboundDone() ? CLOSED : OK, bytesConsumed, bytesProduced, status);
-                            default:
-                                return sslReadErrorResult(SSL.getLastErrorNumber(), bytesConsumed, bytesProduced);
+
+                        int bytesRead = readPlaintextData(dst);
+
+                        if (bytesRead > 0) {
+                            bytesProduced += bytesRead;
+
+                            if (!dst.hasRemaining()) {
+                                // Move to the next dst buffer as this one is full.
+                                dstsOffset++;
+                            } else if (packetLength == 0) {
+                                // We read everything return now.
+                                return newResult(
+                                        isInboundDone() ? CLOSED : OK, bytesConsumed, bytesProduced, status);
+                            } else {
+                                // try to write again to the BIO, so stop reading from it by break out of the readLoop.
+                                break;
+                            }
+                        } else {
+                            int sslError = SSL.getError(ssl, bytesRead);
+                            switch (sslError) {
+                                case SSL.SSL_ERROR_WANT_WRITE:
+                                case SSL.SSL_ERROR_WANT_READ:
+                                    // break to the outer loop as we want to read more data which means we need to write
+                                    // more to the BIO.
+                                    break readLoop;
+
+                                case SSL.SSL_ERROR_ZERO_RETURN:
+                                    // This means the connection was shutdown correctly, close inbound and outbound
+                                    if (!receivedShutdown) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)